src/main/java/seedu/address/model/cardtag/CardTag.java

```diff
@@ -1,111 +1,150 @@
 import java.util.List;
+import java.util.Objects;
 import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
-import com.google.common.graph.GraphBuilder;
-import com.google.common.graph.MutableGraph;
-
+import javafx.collections.ObservableList;
 import seedu.address.model.card.Card;
 import seedu.address.model.tag.Tag;
 
 /**
+ * @author jethro
  * This class captures the relations between cards and tags.
- *
  */
 public class CardTag {
-    private MutableGraph<Node> graph;
+    private HashMap<String, Set<String>> cardMap;
+    private HashMap<String, Set<String>> tagMap;
 
     public CardTag() {
-        this.graph = GraphBuilder.undirected().build();
+        this.cardMap = new HashMap<>();
+        this.tagMap = new HashMap<>();
     }
 
-    public void reset() {
-        this.graph = GraphBuilder.undirected().build();
+    public HashMap<String, Set<String>> getCardMap() {
+        return cardMap;
     }
 
-    public MutableGraph<Node> getGraph() {
-        return graph;
+    public HashMap<String, Set<String>> getTagMap() {
+        return tagMap;
     }
 
-    private void addNode(Node node) {
-        this.graph.addNode(node);
+    public void setCardMap(HashMap<String, Set<String>> cardMap) {
+        this.cardMap = cardMap;
     }
 
-    // Aliases to add tags
-    public void addCard(Card card) {
-        addNode(card);
+    public void setTagMap(HashMap<String, Set<String>> tagMap) {
+        this.tagMap = tagMap;
     }
 
     /**
-     * Adds a list of cards to the graph.
-     * @param cards list of cards
+     * Checks if the Card and Tag given are connected by an edge.
+     * @param cardId UUID of card
+     * @param tagId UUID of tag
+     * @return true if cord and tag are connected, false otherwise
      */
-    public void addCards(List<Card> cards) {
-        for (Card card : cards) {
-            addNode(card);
-        }
+    public boolean isConnected(String cardId, String tagId) {
+        Set<String> tags = cardMap.get(cardId);
+        Set<String> cards = tagMap.get(tagId);
+
+        return (tags != null && tags.contains(tagId))
+                || (cards != null && cards.contains(cardId)); // should always short-circuit here
     }
 
-    /**
-     * Adds a single tag to the graph.
-     * @param tag Tag to add.
-     */
-    public void addTag(Tag tag) {
-        addNode(tag);
+    public boolean isConnected(Card card, Tag tag) {
+        return isConnected(card.getId().toString(), tag.getId().toString());
     }
 
-    /**
-     * Adds a list of tags to the graph.
-     * @param tags list of tags
-     */
-    public void addTags(List<Tag> tags) {
-        for (Tag tag : tags) {
-            addNode(tag);
+
+    public List<Card> getCards(String tagId, ObservableList<Card> cardList) {
+        Set<String> cards = tagMap.get(tagId);
+        if (cards != null) {
+            return cardList.filtered(card -> cards.contains(card.getId().toString()));
```

Comment:
just checking, is it appropriate for predicates to be used in the CardTag class?

Reply:
yeah I have it here because it is specific to CardTag, and does not depend on the properties of `UniqueCardList` or `UniqueTagList`